Add pattern for split Log to independent messages

Zoynels opened the pull request:
Add two ways to fix the difference between logs which stores in files and logs in UI:
1. Custom re.pattern which can split whole log into "real" messages
2. Disable deduplication of log lines

closes: #39686 
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Zoynels committed:
Add the pattern for split Log to independent messages

Zoynels committed:
Add Flag to enable/disable deduplication between log messages (another way to store multiline messages)

Zoynels committed:
Add the pattern for split Log to independent messages (fix misspelling)

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -810,6 +810,20 @@ logging:
       example: ~
       default: "[%%(asctime)s] {{%%(filename)s:%%(lineno)d}} %%(levelname)s - %%(message)s"
+    log_format_re_pattern:
+      description: |
```
Can we elaborate a bit about the default pattern (in human readable language)?

Zoynels replied:
As a default pattern for log is: "[%%(asctime)s] {{%%(filename)s:%%(lineno)d}} %%(levelname)s - %%(message)s"
ChatGPT hepled and then I try with: 
* https://regex101.com/r/tK8tLJ/1 - to split
* https://regex101.com/r/tK8tLJ/2 - to match

`(?<=\n)` - find before \n and where next [date]
`(?=` - what patter we should find to split
   `(?:|[ \t]+?)` in tests where chould be som spaces before date (it is not the default pattern), but I want prepare more universal pattern.

`(?:\[.+?\])` -- detect [%%(asctime)s]
`(?:{.+?:[0-9]+})` -- detect {%%(filename)s:%%(lineno)d} - 
`(?:CRITICAL|ERROR|WARNING|INFO|DEBUG)` - detect %%(levelname)s
`(?:.+?)` - and message

`(?:\n|$)`) - stops

when use re.split() - the symbols `(?:` do not incdule, because all bracet includes as split pattern


After some thinking I simplify it to: `(?<=\n)(?=\[.+?\] {.+?:[0-9]+} \w+ - .*?)`
But then we need remove prefix spaces or with such pattern spaces before 
https://github.com/apache/airflow/blob/7fec562987b9ed8b847f8bb80e316b016d4b309d/tests/utils/test_log_handlers.py#L742-L760 

or `(?<=\n)(?=[ \t]*?\[.+?\] {.+?:[0-9]+} \w+ - .*?)` then no changes https://regex101.com/r/tK8tLJ/3

Zoynels committed:
Add Flag to enable/disable deduplication between log messages (another way to store multiline messages)

rename flag_name to pass spellcheck

Zoynels committed:
Add the pattern for split Log to independent messages (rewrite method)

simplify function and pattern
* currently it splits all lines by \n and then check each line that it match pattern (then it start new message)
use re2 package

Zoynels committed:
fix ruff, mupy and trailing-whitespace

Zoynels committed:
fix tests

now we read file as-is, without strip new lines at the end of the file

s3 - we write \n at the end of the file and then we recieve this \n.

if we didn't write \n then we read equal text

Zoynels committed:
fix
